sympy/functions/combinatorial/factorials.py

```diff
@@ -339,13 +339,18 @@ def eval(cls, arg):
                 if arg.is_even:
                     k = arg / 2
                     return 2 ** k * factorial(k)
-
-                return factorial(arg) / factorial2(arg - 1)
+                if arg.is_odd:
+                    return factorial(arg) / factorial2(arg - 1)
+                raise ValueError("argument must be nonnegative integer or odd integer")
 
             if arg.is_even:
                 raise ValueError("argument must be nonnegative or odd")
 
-            return arg * (S.NegativeOne) ** ((1 - arg) / 2) / factorial2(-arg)
+            if arg.is_integer:
```

Comment:
I think this test could be earlier (in the form ``arg.is_Integer`` which is more efficient). Probably this test above
```
            if arg.is_infinite:
                return
```
could be replaced with
```
            if not arg.is_Integer:
                raise ValueError("argument must be an integer")
```
Then this ``if arg.is_integer``, and also the ``if arg.is_odd`` bit above could be removed to simplify the code.